Shooter and Vision

CooperW12 opened the pull request:
Adding shooter subsystem with methods for setting speed and disabling the shooter. Starting to figure out distance calculations on the vision system.

CooperW12 committed:
add shooter

CooperW12 committed:
cleaned code up

🧹🧼

cyrusduong commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -0,0 +1,96 @@
+    double min = SmartDashboard.getNumber("Min Output", 0);
+    
+    // if PID coefficients on SmartDashboard have changed, write new values to controller
+    if((p != kP)) { m_pidController.setP(p); kP = p; }
+    if((i != kI)) { m_pidController.setI(i); kI = i; }
+    if((d != kD)) { m_pidController.setD(d); kD = d; }
+    if((iz != kIz)) { m_pidController.setIZone(iz); kIz = iz; }
+    if((ff != kFF)) { m_pidController.setFF(ff); kFF = ff; }
```
Usually don't double wrap the conditional if you don't need to.

cyrusduong commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    SmartDashboard.putNumber("ProcessVariable", m_encoder.getVelocity());
+
```
Nit: Empty space here.

cyrusduong replied:
Sorry I see the code formatting is to put spaces after/before curly boys. Disregard

cyrusduong commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -0,0 +1,96 @@
+    SmartDashboard.putNumber("Setpoint", setPoint);
+    m_pidController.setReference(setPoint, CANSparkMax.ControlType.kVelocity);
+  }
+  public void disableShooter(){
```
Nit: Newline between methods.

cyrusduong commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -0,0 +1,96 @@
+  public void periodic(){
+    // This method will be called once per scheduler run
+    // read PID coefficients from SmartDashboard
+    double p = SmartDashboard.getNumber("P Gain", 0);
+    double i = SmartDashboard.getNumber("I Gain", 0);
+    double d = SmartDashboard.getNumber("D Gain", 0);
+    double iz = SmartDashboard.getNumber("I Zone", 0);
+    double ff = SmartDashboard.getNumber("Feed Forward", 0);
+    double max = SmartDashboard.getNumber("Max Output", 0);
+    double min = SmartDashboard.getNumber("Min Output", 0);
```
Unlikely to happen, but what happens if SmartDashboard is disconnected or fails to retrieve the numbers?
Would your pidController set all the values to 0?

CooperW12 replied:
These live SmartDashboard numbers are just there for testing so I can change the numbers without changing the code and redeploying so they probably won't be on the final robot code. Great problem scenario though I will keep that in mind when getting numbers from SmartDashboard especially with the vision subsystem since that gets all of its values from the network tables on SmartDashboard. Thanks for the feedback

cyrusduong reviewed (dismissed):
Mostly code formatting/nits. Looks good.

CooperW12 committed:
polishing the shooter code

fixed some stuff based on feedback and added a few comments to clear things up

CooperW12 commented:
Polished the code up a bit. Thanks again for the feedback Cyrus!

Team1710CTO commented on src/main/java/frc/robot/subsystems/ShooterSubsystem.java:
```diff
@@ -0,0 +1,89 @@
+    m_encoder = m_motor.getEncoder();
+
+    // PID coefficients
+    kP = 0.0001; 
```
There should be a preset location in the Constants.Java file for your PID constants. store those values in that file and reference them here, same goes with the motor controller ID

Andrew-K961 reviewed (changes requested):
Hudson makes a good point, the PID constants should be moved to the Constants.java file. Everything else looks good.

CooperW12 committed:
Photon Vision Subsystem almost done

Figured out the distance calculation stuff got it within .10 of a meter with unprecise values in the calculation which is pretty good but ill add the actual values in once I know where the cameras will be on the robot. Also created the methods and doubles that will be called in the commands to get the distance to the goal/ball, set the alliance pipeline, and find the X displacement of the goal. :)

CooperW12 committed:
Update ShooterSubsystem.java

CooperW12 committed:
Finished code until testing

Andrew-K961 reviewed (dismissed):
Looks good

Team1710CTO commented on src/main/java/frc/robot/commands/DefaultDriveCommand.java:
```diff
@@ -4,6 +4,8 @@
 import edu.wpi.first.wpilibj2.command.CommandBase;
 import frc.robot.subsystems.DrivetrainSubsystem;
 import frc.robot.subsystems.GyroSubsystem;
+import frc.robot.subsystems.PhotonVisionSubsystem;
```
do these commands need the vision and shooter subsystem?

CooperW12 committed:
remove import

CooperW12 committed:
Added a ton of comments

👍

Team1710CTO reviewed (approved):
Looks good.
